### ISSUE TITLE
Exception from DMR parser on NaN Float32 attribute (DAP4)

### ISSUE THREAD

**Issue by peterazmanov** [pydap/pydap#344]
### Description

I am getting an exception when trying to access netCDF dataset via xarray/DAP4, the exception is raised during parsing of DMR.

### Minimal example that reproduces the bug

```python
Python 3.12.3 (main, Apr 27 2024, 19:00:21) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pydap.client
>>> data = pydap.client.open_dmr_file("wind.nc.dmr")
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/user/.cache/pypoetry/virtualenvs/test-t6h5nzz--py3.12/lib/python3.12/site-packages/pydap/client.py", line 132, in open_dmr_file
    dataset = pydap.parsers.dmr.dmr_to_dataset(dmr)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.cache/pypoetry/virtualenvs/test-t6h5nzz--py3.12/lib/python3.12/site-packages/pydap/parsers/dmr.py", line 151, in dmr_to_dataset
    variable["attributes"] = get_attributes(variable["element"], {})
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.cache/pypoetry/virtualenvs/test-t6h5nzz--py3.12/lib/python3.12/site-packages/pydap/parsers/dmr.py", line 91, in get_attributes
    value = ast.literal_eval(value)
            ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/ast.py", line 112, in literal_eval
    return _convert(node_or_string)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/ast.py", line 111, in _convert
    return _convert_signed_num(node)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/ast.py", line 85, in _convert_signed_num
    return _convert_num(node)
           ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/ast.py", line 76, in _convert_num
    _raise_malformed_node(node)
  File "/usr/lib/python3.12/ast.py", line 73, in _raise_malformed_node
    raise ValueError(msg + f': {node!r}')
ValueError: malformed node or string on line 1: <ast.Name object at 0x7f7be3b680d0>
```
Version:
```
pydap @ git+https://github.com/pydap/pydap.git@742af30ae32d0b6d773f6d2c2568a98473d180e4
```
### Details, possible fix

DMR atomic types are parsed with `ast.literal_eval`:
https://github.com/pydap/pydap/blob/1e76c1987b72192b9664418674d83e73a57c2bda/src/pydap/parsers/dmr.py#L83-L93

It's not correct since the specification allows "inf" and "NaN" constants which are not valid Python expressions (http://docs.opendap.org/index.php/DAP4:_Specification_Volume_1#Atomic_Types).

I tried the following quick fix:
```diff
         name = attribute_element.get("name")
         value = attribute_element.find("Value").text
         _type = attribute_element.get("type")
-        if _type in numType:
+        if _type in {"Float32", "Float64"}:
+            value = float(value)
+        elif _type in numType:
             value = ast.literal_eval(value)
         attributes[name] = value
     return attributes
```

... and it worked:
```python
Python 3.12.3 (main, Apr 27 2024, 19:00:21) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pydap.client
>>> data = pydap.client.open_dmr_file("wind.nc.dmr")
>>> data["speed_wind_true_u"].attributes
{'units': 'kn', 'missing_value': nan}
```

<details>

<summary>wind.nc.dmr</summary>

```xml
<?xml version="1.0" encoding="ISO-8859-1"?>
<Dataset xmlns="http://xml.opendap.org/ns/DAP/4.0#" xml:base="https://internal/opendap/hyrax/wind.nc" dapVersion="4.0" dmrVersion="1.0" name="wind.nc">
    <Dimension name="time" size="65"/>
    <Dimension name="latitude" size="181"/>
    <Dimension name="longitude" size="360"/>
    <Float32 name="time">
        <Dim name="/time"/>
        <Attribute name="units" type="String">
            <Value>hour since 2024-07-10T00:00:00Z</Value>
        </Attribute>
    </Float32>
    <Float32 name="latitude">
        <Dim name="/latitude"/>
        <Attribute name="units" type="String">
            <Value>degrees_north</Value>
        </Attribute>
    </Float32>
    <Float32 name="longitude">
        <Dim name="/longitude"/>
        <Attribute name="units" type="String">
            <Value>degrees_east</Value>
        </Attribute>
    </Float32>
    <Float32 name="speed_wind_true_u">
        <Dim name="/time"/>
        <Dim name="/latitude"/>
        <Dim name="/longitude"/>
        <Attribute name="units" type="String">
            <Value>kn</Value>
        </Attribute>
        <Attribute name="missing_value" type="Float32">
            <Value>NaN</Value>
        </Attribute>
        <Map name="/time"/>
        <Map name="/latitude"/>
        <Map name="/longitude"/>
    </Float32>
    <Float32 name="speed_wind_true_v">
        <Dim name="/time"/>
        <Dim name="/latitude"/>
        <Dim name="/longitude"/>
        <Attribute name="units" type="String">
            <Value>kn</Value>
        </Attribute>
        <Attribute name="missing_value" type="Float32">
            <Value>NaN</Value>
        </Attribute>
        <Map name="/time"/>
        <Map name="/latitude"/>
        <Map name="/longitude"/>
    </Float32>
</Dataset>
```

</details>

**Comment by ndp-opendap** [comment]
Python (and numpy and pandas) have built in support for NaN values. We should utilize it here:
https://www.turing.com/kb/nan-values-in-python#

Here is a built in comparison example from that article:
```
value = float('nan') 
if value != value: 
    print("Value is NaN") 
else: print("Value is not NaN")
```

**Comment by Mikejmnez** [comment]
Sorry it took me some time. Here are my thoughts:

1. This is a great opportunity to catch and pass useful errors down to, say `xarray`. 
2. When parsing the DMR, values of attributes defined as `Float32` and other types are extracted frustratingly as `<class 'str'>`, which in python would violate CF-conventions. And so `ast.literal_eval` allows to get numbers (ints, floats) back. It may strictly violate CF: an attribute on the DMR with value `-9999` and type `Float32` would pass as an `int`, but I consider this scenario a mild violation...

And so without re-writing all the code previous done to parse the DMR, I found this a potential solution:

https://github.com/pydap/pydap/commit/8a3e71f857f6ef8a07593437064b8110b28bec42


The changes then:
1. Check if there are NaNs or infs when there is a ValueError Exception (the kind you got when dealing with NaN vals)
2. I checked, it works with `NaN`, `nan`, `INF`, `Inf`, `inf`.

And it works with your example
```python
ds = open_dmr_file("wind.nc.dmr")
ds.tree()
.wind.nc
├──time
├──latitude
├──longitude
├──speed_wind_true_u
└──speed_wind_true_v

ds['speed_wind_true_u'].attributes
{'units': 'kn', 'missing_value': nan}
```

**Comment by peterazmanov** [comment]
Thanks for the quick response!

The new version seems more robust than the previous one in the sense that it can parse more literals. However, looking at the grammar I see some allowed literals that can raise the error, perhaps since the parser in its current state is not compliant, it's safer to issue a warning and leave the value as `str` in case of unexpected input.

P.S. IMHO `ast.literal_eval` allows some nonsense like getting `[]` from Float32 attribute, which is why I put the `float(value)` conversion first. Of course, that wouldn't get rid of `ast.literal_eval` entirely.

**Comment by Mikejmnez** [comment]
Thanks @peterazmanov I like your suggestions: raise a `Warning` instead of `TypeError`, and move first the `float evaluation` in the conditionals before the `ast.literal_eval` (which is now the last resort). I ended up also including an option to distinguish between integers and floats (and so a bit more `CF-compliant`).

**Comment by Mikejmnez** [comment]
closed by #345

**Comment by peterazmanov** [comment]
By "issue a warning" I meant `warnings.warn`, which by default only shows the warning, it gives client apps more control, they can silence selected warnings or turn them into exceptions.
